troposphere/haze/include/haze/ptp_data_builder.hpp

```diff
@@ -0,0 +1,164 @@
+                };
+
+                if (m_disabled) {
+                    R_SUCCEED();
+                } else {
+                    R_RETURN(m_server->WritePacket(m_data, m_offset));
+                }
+            }
+        public:
+            constexpr explicit PtpDataBuilder(void *data, AsyncUsbServer *server) : m_server(server),  m_transmitted_size(), m_offset(), m_data(static_cast<u8 *>(data)), m_disabled() { /* ... */ }
+
+            Result Commit() {
+                if (m_offset > 0) {
+                    /* If there is remaining data left to write, write it now. */
+                    R_TRY(this->Flush());
+                }
+
+                if (m_transmitted_size % PtpUsbBulkHighSpeedMaxPacketLength == 0) {
+                    /* If the transmission size was a multiple of wMaxPacketSize, send a zero length packet. */
+                    R_TRY(this->Flush());
+                }
+
+                R_SUCCEED();
+            }
+
+            Result AddBuffer(const u8 *buffer, u32 count) {
+                while (count > 0) {
+                    /* Calculate how many bytes we can write now. */
+                    u32 write_size = std::min<u32>(count, haze::UsbBulkPacketBufferSize - m_offset);
+
+                    /* Write this number of bytes. */
+                    std::memcpy(m_data + m_offset, buffer, write_size);
+                    m_offset += write_size;
+                    buffer += write_size;
+                    count -= write_size;
+
+                    /* If we cannot write more bytes now, flush. */
+                    if (m_offset == haze::UsbBulkPacketBufferSize) {
+                        R_TRY(this->Flush());
+                    }
+                }
+
+                R_SUCCEED();
+            }
+
+            template <typename T>
+            Result Add(T value) {
+                const auto bytes = std::bit_cast<std::array<u8, sizeof(T)>>(value);
+
+                R_RETURN(this->AddBuffer(bytes.data(), bytes.size()));
+            }
+
+            Result AddDataHeader(PtpUsbBulkContainer &request, u32 data_size) {
+                R_TRY(this->Add<u32>(PtpUsbBulkHeaderLength + data_size));
+                R_TRY(this->Add<u16>(PtpUsbBulkContainerType_Data));
+                R_TRY(this->Add<u16>(request.code));
+                R_TRY(this->Add<u32>(request.trans_id));
+
+                R_SUCCEED();
+            }
+
+            Result AddResponseHeader(PtpUsbBulkContainer &request, PtpResponseCode code, u32 params_size) {
+                R_TRY(this->Add<u32>(PtpUsbBulkHeaderLength + params_size));
+                R_TRY(this->Add<u16>(PtpUsbBulkContainerType_Response));
+                R_TRY(this->Add<u16>(code));
+                R_TRY(this->Add<u32>(request.trans_id));
+
+                R_SUCCEED();
+            }
+
+            template <typename F>
+            Result WriteVariableLengthData(PtpUsbBulkContainer &request, F &&func) {
+                m_disabled = true;
```

Comment:
The reason this exists is basically to compute the size of the response length, as we don't know how big it's going to be in advance (and computing it in the same pass could theoretically require extra memory, if our buffers were smaller - think 512 bytes, instead of 1 MiB).

I have refactored it a bit to make the scoping explicit - let me know what you think.